folding-schemes/src/commitment/ipa.rs

```diff
@@ -0,0 +1,693 @@
+    pub fn verify(
+        params: &PedersenParams<C>,
+        transcript: &mut impl Transcript<C>,
+        x: &C::ScalarField, // evaluation point
+        v: &C::ScalarField, // value at evaluation point
+        P: &C,              // commitment
+        p: &Proof<C>,
+        r: &C::ScalarField, // blinding factor
+        k: usize,           // k = log2(d), where d is the degree of the committed polynomial
+    ) -> Result<bool, Error> {
+        if !H && (!p.l.is_empty() || !p.r.is_empty()) {
+            return Err(Error::CommitmentVerificationFail);
+        }
+        if H && (p.l.len() != k || p.r.len() != k) {
+            return Err(Error::CommitmentVerificationFail);
+        }
+        if p.L.len() != k || p.R.len() != k || p.u_invs.len() != k {
+            return Err(Error::CommitmentVerificationFail);
+        }
+
+        // absorbs & get challenges
+        transcript.absorb_point(P)?;
+        let s = transcript.get_challenge();
+        let U = C::generator().mul(s);
+        let mut u: Vec<C::ScalarField> = vec![C::ScalarField::zero(); k];
+        for i in (0..k).rev() {
+            transcript.absorb_point(&p.L[i])?;
+            transcript.absorb_point(&p.R[i])?;
+            u[i] = transcript.get_challenge();
+        }
+
+        let P = *P + U.mul(v);
+
+        let mut q_0 = P;
+        let mut r = *r;
+
+        // check correctnes of the u_invs delegated to the prover
+        for (j, u_j) in u.iter().enumerate() {
+            if *u_j * p.u_invs[j] != C::ScalarField::one() {
+                return Err(Error::CommitmentVerificationFail);
+            }
+        }
+
+        // compute b & G from s
+        let s = build_s(&u, &p.u_invs, k)?;
+        // b = <s, b_vec> = <s, [1, x, x^2, ..., x^d-1]>
+        let b = s_b_inner(&u, x)?;
+        let d: usize = 2_u64.pow(k as u32) as usize;
+        if params.generators.len() < d {
+            return Err(Error::PedersenParamsLen(params.generators.len(), d));
+        }
+        let G = C::msm_unchecked(&params.generators, &s);
+
+        for (j, u_j) in u.iter().enumerate() {
+            let uj2 = u_j.square();
+            let uj_inv2 = p.u_invs[j].square();
+
+            q_0 = q_0 + p.L[j].mul(uj2) + p.R[j].mul(uj_inv2);
+            if H {
+                r = r + p.l[j] * uj2 + p.r[j] * uj_inv2;
+            }
+        }
+
+        let q_1 = if H {
+            G.mul(p.a) + params.h.mul(r) + U.mul(p.a * b)
+        } else {
+            G.mul(p.a) + U.mul(p.a * b)
+        };
+
+        Ok(q_0 == q_1)
+    }
+}
+
+// s = (
+//   u₁⁻¹ u₂⁻¹ … uₖ⁻¹,
+//   u₁   u₂⁻¹ … uₖ⁻¹,
+//   u₁⁻¹ u₂   … uₖ⁻¹,
+//   u₁   u₂   … uₖ⁻¹,
+//   ⋮    ⋮      ⋮
+//   u₁   u₂   … uₖ
+// )
+// naively would take k * d = k * 2^k time, with the current approach taking advantadge of the
+// symmetric nature of s, it takes k * d/2 = k * (2^k)/2 = k * 2^{k-1} time.
+fn build_s<F: PrimeField>(u: &[F], u_invs: &[F], k: usize) -> Result<Vec<F>, Error> {
```

Comment:
Inspired by [IPA in halo2](https://github.com/privacy-scaling-explorations/halo2/blob/main/halo2_backend/src/poly/ipa/strategy.rs#L156), We could have less multiplications here using DP like:

```rust
    let d: usize = 2_u64.pow(k as u32) as usize;
    let mut s: Vec<F> = vec![F::one(); d];
    for (len, (u_j, u_j_inv)) in u
        .iter()
        .zip(u_invs)
        .enumerate()
        .map(|(i, u_j)| (1 << i, u_j))
    {
        let (left, right) = s.split_at_mut(len);
        let right = &mut right[0..len];
        right.copy_from_slice(left);
        for s in left {
            *s *= u_j_inv;
        }
        for s in right {
            *s *= u_j;
        }
    }
```

It only needs <code>2<sup>k+1</sup> - 2</code> instead of <code>(k+1)2<sup>k-1</sup></code> multiplications, and same for the gadget below.

Reply:
Wow! This is great! Thank you very much for pointing this! The reduction in the number of constraints is considerable!
The verification circuit now takes
- for k=4: -9k constraints (-7%)
- for k=8: -473k constr (-31%)
- for k=9: -1123k constr (-35%)
- for k=10: -2578k constr (-39%)

And now IPA verification (without amortizing) is very close to Pedersen verification (in-circuits).
Again, thank you very much! I always learn new things from your reviews :raised_hands: